### PR TITLE
Add support for linting a subset of files in the package

### LINKED ISSUE #80: Running custom_lint on a single file
Thanks for this awesome plugin. I recently integrated in my org's project and the custom lint is working like charm. Uses quite some memory but that's a different issue.
I wanted to ask is there any way to lint a single file ? Like from command line ? 

**Describe what scenario you think is uncovered by the existing examples/articles**

Currently, following the documentation, I am able to run `flutter pub run custom_lint` in the project root and it scans all the files of the project. I couldn't find any documentation which states how to lint a single file from command line.

**Additional context**
We are working on a team project in an org and want to add Documentation compliance for new files. As all the existing files can't be documented at once, we want that any new PR should include complete documentation for the changed files. 
If it was possible to run `custom_lint` on a single file, then this task could easily be integrated to GitHub workflow of our project.

### PR DESCRIPTION
**Pull request by literalEval**
Fixes #80 

The `custom_lint` cli tool is great and all, but lacks a serious feature of scanning and providing errors/warning of only a given list of files/folders. This PR aims to add that feature to `custom_lint`.

Changes - 

1. List of files of your package that you want to scan.
2. Directories of your package that you want to scan. `custom_lint` will recursively find all the files in the given directories and scan them. 
3. These paths can be given in any order and can be both relative and absolute.

### CODE REVIEW

**Comment by literalEval** [comment]
@rrousselGit please review and merge

**Comment by palisadoes** [comment]
@rrousselGit This is a really important feature addition to our repo. It will allow us to individually check files that were modified in a pull requests.

We have been having difficulties in maintaining consistent in parts of our code base. Rather than doing a 100% refactoring, this will allow us to incrementally enforce our best practices.

It's usefulness is great to us. Any assistance in reviewing would be welcome.

**Comment by rrousselGit** [comment]
This is missing tests. Could you add tests for all that were added?

File: packages/custom_lint/bin/custom_lint.dart [hunk 1]
```diff
@@ -26,9 +27,53 @@ Future<void> entrypoint([List<String> args = const []]) async {
...
+      final dir = Directory(restArgs[i]);
+      final fileEntities = dir.listSync(recursive: true);
+
+      for (final fileEntity in fileEntities) {
+        fileList.add(fileEntity.path);
+      }
+    }
+  }
+
+  // for (var i = 0; i < dirList.length; i++) {
```
Comment (rrousselGit) [inline packages/custom_lint/bin/custom_lint.dart:57]: Can you remove this? :)

File: packages/custom_lint/lib/src/analyzer_plugin_starter.dart [hunk 2]
```diff
@@ -5,11 +5,12 @@ import 'plugin_delegate.dart';
 import 'v2/custom_lint_analyzer_plugin.dart';
 
 /// Connects custom_lint to the analyzer server using the analyzer_plugin protocol
-void start(Iterable<String> _, SendPort sendPort) {
+void start(Iterable<String> _, SendPort sendPort, List<String>? fileList) {
```
Comment (rrousselGit) [inline packages/custom_lint/lib/src/analyzer_plugin_starter.dart:8]: Why is this modified? This shouldn't be needed by the CLI

File: packages/custom_lint_builder/lib/src/client.dart [hunk 3]
```diff
@@ -655,7 +659,7 @@ class _ClientAnalyzerPlugin extends ServerPlugin {
     // We customize the behavior to optimize analysis to be performed only
     // if something changed
     if (paths.isEmpty) return Future.value();
-
+
```
Comment (rrousselGit) [inline packages/custom_lint_builder/lib/src/client.dart:662]: This isn't formatted properly

File: packages/custom_lint_builder/lib/src/client.dart [hunk 4]
```diff
@@ -103,6 +104,9 @@ class CustomLintPluginClient {
   /// Whether the client should perform a hot-reload when the source of plugins changes.
   final bool watchMode;
 
+  /// List of files to lint
+  List<String>? fileList;
```
Comment (rrousselGit) [inline packages/custom_lint_builder/lib/src/client.dart:108]: should be final

File: packages/custom_lint/test/run_plugin.dart [hunk 5]
```diff
@@ -10,14 +10,16 @@ import 'package:path/path.dart' as p;
...
   Directory directory,
   // to ignoreErrors as we cannot explictly handle errors
 ) async {
-  final runner = startRunnerForApp(directory, includeBuiltInLints: false);
+  final runner = startRunnerForApp(fileList, directory, includeBuiltInLints: false);
   return runner.getLints(reload: false);
 }
 
 CustomLintRunner startRunnerForApp(
+  List<String>? fileList,
```
Comment (rrousselGit) [inline packages/custom_lint/test/run_plugin.dart:22]: Can you make this a named parameter?

File: packages/custom_lint/test/run_plugin.dart [hunk 6]
```diff
@@ -10,14 +10,16 @@ import 'package:path/path.dart' as p;
 import 'package:test/test.dart';
 
 Future<List<AnalysisErrorsParams>> runServerInCliModeForApp(
+  List<String>? fileList,
```
Comment (rrousselGit) [inline packages/custom_lint/test/run_plugin.dart:13]: Make this a named parameter

File: packages/custom_lint/bin/custom_lint.dart (hunk 1 above)
Comment (literalEval) [reply packages/custom_lint/bin/custom_lint.dart]: Very sorry for this. Was so excited, just didn't look at that properly :)

**Comment by literalEval** [comment]
@rrousselGit please look again. 
And about the test, I'll add them once this gets merged.

**Comment by rrousselGit** [comment]
> @rrousselGit please look again. And about the test, I'll add them once this gets merged.

That's not how it works :)

**Comment by literalEval** [comment]
Okay then @rrousselGit. My exams are going on, so I'll try to get some time and do it. 
Meanwhile, does the changed code look good to you ? And also, all I would need to do is to add a test that shows lint warnings of a specific set of files, given a lot of files in the package. Right ?

**Comment by rrousselGit** [comment]
Eveything that's present needs to be tested. So this includes the relative/absolute check, the file vs folder & possibly more.

We would also need to test that this works with 2+ packages. For example `custom_lint packages/foo/foo.dart packages/bar/bar.dart`

We'd also want to check that packages not listed in the file/directory list aren't considered. So when doing `custom_lint packages/foo`, if there is a `packages/bar` project and it contains a plugin which `packages/foo` doesn't uses, the CLI shouldn't use said plugin.

**Comment by literalEval** [comment]
@rrousselGit Yeah the folder and relative etc need to be tested. Got it.

But about the other things like multiple plugin, no functionality has been changed regarding that, so is any test needed ? Like this PR only says that if you have a `package` then now you can analyze specific files from it. It doesn't change anything regarding which plugin should be loaded. So the existing tests would be good for them right ?

**Comment by rrousselGit** [comment]
@literalEval It should though

The current implementation checks a lot of files that are unnecessary.  You should modify `CustomLintRunner` in `runner.dart` to have `_allContextRoots` include list the context roots associated with searched files.

**Comment by rrousselGit** [comment]
Also we would need a test where a file path points to a file that doesn't exist. And one where a file doesn't have custom_lint enabled

**Comment by literalEval** [comment]
Okay. I'll check and let you know.

**Comment by literalEval** [comment]
@rrousselGit closing this. Will start working on it again from scratch :)
